### PR TITLE
feat(hardfork): ckb2021 hardfork features

### PR DESCRIPTION
**Pull request by yangby-cryptape**
### Suggestions for Review

**The best way to review this PR is reviewing each commit independently.**

#### New hardfork Features

**All following features can be found in RFC PRs.**

- feat(hardfork): in the "since epoch", the index should be less than length

  Ref: [CKB-RFCs PR 223: Ensure that index < length in input since field using epoch](https://github.com/nervosnetwork/rfcs/pull/223)

- feat(hardfork): use block timestamp of input cells as relative since start timestamp

  Ref: [CKB-RFCs PR 221: Use Block Timestamp as Start Timestamp in Since](https://github.com/nervosnetwork/rfcs/pull/221)

- **[Reverted]** ~~feat(hardfork): allow unknown block versions and transactions versions~~

  ~~Ref: [CKB-RFCs PR 230: Allow unknown tx & block version](https://github.com/nervosnetwork/rfcs/pull/230)~~

- feat(hardfork): allow script multiple matches on identical data for type hash-type scripts

  Ref: [CKB-RFCs PR 222: Allow script multiple matches on identical code](https://github.com/nervosnetwork/rfcs/pull/222)

- feat(hardfork): reuse the uncles hash in the header as the extra hash

  Ref: [CKB-RFCs PR 224: Add a variable length field in the block header](https://github.com/nervosnetwork/rfcs/pull/224)

- **[Reverted]** ~~feat(hardfork): allow loading uncommitted cell data hashes from tx pool~~

  ~~Ref: [CKB-RFCs PR 228: ckb2021: fix load_cell_data_hash syscall](https://github.com/nervosnetwork/rfcs/pull/228)~~

#### Other Important Commits

- feat(hardfork): setup the components for hard fork features

- refactor: let verifiers know the real environment that the transaction is in

  Almost all features require this refactor commit.

- refactor: remove useless parameter "with_data" because it always be true (tricky)

  So I can change less APIs and less code to apply the feature: allow loading uncommitted cell data hashes from tx pool.

### About Tests

Almost all features have detailed integration tests (or unit tests):
- Many blocks before hardfork;
- Only one block before hardfork;
- The block at hardfork;
- Many blocks after hardfork.

All commits can passed all integration tests and unit tests.

### CODE REVIEW

**Comment by yangby-cryptape** [comment]
Hold until the CI checks are fixed.

**Comment by quake** [comment]
benchmark

**Comment by quake** [comment]
sync-test

**Comment by yangby-cryptape** [comment]
> #### Benchmark Result
>   - TPS: 494
>   - CKB Version: fb0fbea2c6fdd4dd848be8b005da73932c76fb71

Not bad. Same as the previous.

**Comment by nervos-bot-user** [comment]
#### Benchmark Result
  - TPS: 486
  - CKB Version: fb0fbea2c6fdd4dd848be8b005da73932c76fb71
  - Instance Type: c5.xlarge
  - Instances Count: 3
  - Bench Type: In2Out2
  - CKB Logger Filter: info,ckb=debug

**Comment by nervos-bot-user** [comment]
#### Sync Test Result
  - CKB Version: fb0fbea2c6fdd4dd848be8b005da73932c76fb71
  - Instance Region: ap-southeast-1
  - Sync Start: 2021-05-26 12:15:50
  - Sync End: 2021-05-26 14:14:52
  - Sync Height: 4403249
  - Sync Speed: 616 blocks/s

**Comment by yangby-cryptape** [comment]
> #### Sync Test Result
>   - CKB Version: fb0fbea2c6fdd4dd848be8b005da73932c76fb71
>   - Instance Region: ap-southeast-1
>   - Sync Height: 4403249
>   - Sync Speed: 616 blocks/s

Better than the average of last 20 tests.

**Comment by yangby-cryptape** [comment]
benchmark

**Comment by yangby-cryptape** [comment]
sync-test

**Comment by nervos-bot-user** [comment]
#### Benchmark Result
  - TPS: 504
  - CKB Version: 0316e0ecb2823f91fdd9bd5fc62d07ca94c87498
  - Instance Type: c5.xlarge
  - Instances Count: 3
  - Bench Type: In2Out2
  - CKB Logger Filter: info,ckb=debug

**Comment by yangby-cryptape** [comment]
> TPS: 504
> CKB Version: 0316e0e

The 3rd best in last 30 tests on branch develop.

**Comment by nervos-bot-user** [comment]
#### Sync Test Result
  - CKB Version: 0316e0ecb2823f91fdd9bd5fc62d07ca94c87498
  - Instance Region: ap-southeast-1
  - Sync Start: 2021-05-31 06:23:39
  - Sync End: 2021-05-31 08:55:24
  - Sync Height: 4440264
  - Sync Speed: 487 blocks/s

**Comment by yangby-cryptape** [comment]
sync-test

**Comment by nervos-bot-user** [comment]
#### Sync Test Result
  - CKB Version: 0316e0ecb2823f91fdd9bd5fc62d07ca94c87498
  - Instance Region: ap-southeast-1
  - Sync Start: 2021-05-31 09:36:34
  - Sync End: 2021-05-31 11:37:47
  - Sync Height: 4441190
  - Sync Speed: 610 blocks/s

**Comment by yangby-cryptape** [comment]
>   CKB Version: 0316e0e
>   Sync Speed: 616 blocks/s

The 3rd best in last 20 tests on branch develop.

File: verification/src/transaction_verifier.rs [hunk 1]
```diff
@@ -693,9 +707,19 @@ impl<'a, DL: HeaderProvider> SinceVerifier<'a, DL> {
...
+                    let parent_hash = self.tx_env.parent_hash();
+                    let epoch_number = self.tx_env.epoch_number(proposal_window);
+                    let hardfork_switch = self.consensus.hardfork_switch();
+                    let base_timestamp = if hardfork_switch
+                        .is_block_ts_as_relative_since_start_enabled(epoch_number)
+                    {
+                        self.parent_median_time(&info.block_hash)
+                    } else {
+                        self.parent_block_time(&info.block_hash)
+                    };
```
Comment (yangby-cryptape) [inline verification/src/transaction_verifier.rs:720]: Made a mistake after rebase.

```suggestion
                    let base_timestamp = if hardfork_switch
                        .is_block_ts_as_relative_since_start_enabled(epoch_number)
                    {
                        self.parent_block_time(&info.block_hash)
                    } else {
                        self.parent_median_time(&info.block_hash)
                    };
```

File: tx-pool/src/component/pending.rs [hunk 2]
```diff
@@ -84,7 +84,7 @@ impl CellProvider for PendingQueue {
                     let mut cell_meta = CellMetaBuilder::from_cell_output(output, data)
                         .out_point(out_point.to_owned())
                         .build();
-                    if !with_data {
+                    if !allow_in_txpool && !with_data {
```
Comment (quake) [inline tx-pool/src/component/pending.rs:87]: This line looks like changed the original behavior, I think the original code is wrong, should be:
```
                    if !with_data {
                        cell_meta.mem_cell_data_hash = None;
                    }
``

**Comment by yangby-cryptape** [comment]
I will apply all request changes at last in one commit.

File: verification/src/transaction_verifier.rs [hunk 3]
```diff
@@ -622,14 +588,20 @@ impl<'a, DL: HeaderProvider> SinceVerifier<'a, DL> {
         self.block_median_time(&parent_hash)
     }
 
+    fn parent_block_time(&self, block_hash: &Byte32) -> u64 {
```
Comment (quake) [inline verification/src/transaction_verifier.rs:591]: This fn name can be easily misinterpreted as getting the timestamp of the parent block, suggest to remove this fn, we can use the inline call `self.data_loader.get_header(block_hash).expect().timestamp()` directly

**Review by quake — APPROVED** [review]

**Review by quake — APPROVED** [review]

**Review by quake — APPROVED** [review]

**Review by quake — APPROVED** [review]

**Review by zhangsoledad — APPROVED** [review]

**Review by doitian — APPROVED** [review]
Let bors work

**Comment by quake** [comment]
bors r=quake,doitian,zhangsoledad

**Pull request merged**